lib/src/client.dart

```diff
@@ -396,31 +399,40 @@ class ClientMethod<Q, R> {
   ClientMethod(this.path, this.requestSerializer, this.responseDeserializer);
 }
 
+/// Provides per-RPC metadata.
+///
+/// Metadata providers will be invoked for every RPC, and can add their own
+/// metadata to the RPC. If the function returns a [Future], the RPC will await
+/// completion of the returned [Future] before transmitting the request.
+typedef FutureOr<Null> MetadataProvider(Map<String, String> metadata);
+
 /// Runtime options for an RPC.
 class CallOptions {
   final Map<String, String> metadata;
   final Duration timeout;
+  final List<MetadataProvider> metadataProviders;
 
-  CallOptions._(this.metadata, this.timeout);
+  CallOptions._(this.metadata, this.timeout, this.metadataProviders);
 
-  factory CallOptions({Map<String, String> metadata, Duration timeout}) {
-    final sanitizedMetadata = <String, String>{};
-    metadata?.forEach((key, value) {
-      final lowerCaseKey = key.toLowerCase();
-      if (!lowerCaseKey.startsWith(':') &&
-          !_reservedHeaders.contains(lowerCaseKey)) {
-        sanitizedMetadata[lowerCaseKey] = value;
-      }
-    });
-    return new CallOptions._(new Map.unmodifiable(sanitizedMetadata), timeout);
+  factory CallOptions(
+      {Map<String, String> metadata,
+      Duration timeout,
+      List<MetadataProvider> providers}) {
```

Comment:
You may want to document here that the `MetadataProviders` are invoked with the given `metadata` map in turn and may each modify it to their liking. Last `MetadataProvider` decides the outcome (modulo sanitization).